modules/peridynamics/src/mesh/PeridynamicsMesh.C

```diff
@@ -0,0 +1,484 @@
+
+  if (_has_cracks)
+  {
+    if (isParamValid("cracks_width"))
+    {
+      _cracks_width = getParam<std::vector<Real>>("cracks_width");
+      if (_cracks_width.size() != _cracks_start.size())
+        mooseError("Number of cracks width is NOT the same as number of cracks!");
+    }
+    else
+    {
+      for (unsigned int i = 0; i < _cracks_start.size(); ++i)
+        _cracks_width.push_back(0);
+    }
+  }
+  else
+    _cracks_width.push_back(0);
+}
+
+PeridynamicsMesh::~PeridynamicsMesh() { _node_coord.clear(); }
+
+std::unique_ptr<MooseMesh>
+PeridynamicsMesh::safeClone() const
+{
+  return libmesh_make_unique<PeridynamicsMesh>(*this);
+}
+
+void
+PeridynamicsMesh::buildMesh()
+{
+  _mesh = _app.getMeshGeneratorMesh();
+  _mesh->prepare_for_use(/*skip_renumber =*/true);
+}
+
+unsigned int
+PeridynamicsMesh::dimension() const
+{
+  return _dim;
+}
+
+dof_id_type
+PeridynamicsMesh::nPDNodes() const
+{
+  return _total_nodes;
+}
+
+dof_id_type
+PeridynamicsMesh::nPDElems() const
+{
+  return _total_bonds;
+}
+
+void
+PeridynamicsMesh::createExtraPeridynamicsMeshData(MeshBase & fe_mesh)
+{
+  // create unconventional mesh data for peridynamics mesh
+  _dim = fe_mesh.mesh_dimension();
+  _total_nodes = fe_mesh.n_elem();
+
+  _node_coord.resize(_total_nodes);
+  _node_mesh_spacing.resize(_total_nodes);
+  _node_horizon.resize(_total_nodes);
+  _node_vol.resize(_total_nodes);
+  _node_horizon_vol.resize(_total_nodes);
+  _node_blockID.resize(_total_nodes);
+  _horizon_neighbors.resize(_total_nodes);
+  _node_associated_bonds.resize(_total_nodes);
+  _bah_dgneighbors.resize(_total_nodes);
+  _bah_vol.resize(_total_nodes);
+  _bah_vol_sum.resize(_total_nodes);
+
+  Real dist = 0.0;
+  // loop through all fe elements to generate PD nodes structure
+  for (MeshBase::element_iterator it = fe_mesh.elements_begin(); it != fe_mesh.elements_end(); ++it)
+  {
+    Elem * fe_elem = *it;
+    // calculate the mesh spacing as average distance between fe element with its neighbors
+    unsigned int nneighbors = 0;
+    Real dist_sum = 0.0;
+    for (unsigned int i = 0; i < fe_elem->n_neighbors(); ++i)
+      if (fe_elem->neighbor_ptr(i) != NULL)
+      {
+        dist = (fe_elem->centroid() - fe_elem->neighbor_ptr(i)->centroid()).norm();
+        dist_sum += dist;
+        nneighbors += 1;
+      }
+    _node_coord[fe_elem->id()] = fe_elem->centroid();
+    _node_mesh_spacing[fe_elem->id()] = dist_sum / nneighbors;
+    _node_horizon[fe_elem->id()] =
+        (_has_horizon_number ? _horizon_number * dist_sum / nneighbors : _horizon_radius);
+    _node_vol[fe_elem->id()] = fe_elem->volume();
+    _node_horizon_vol[fe_elem->id()] = 0.0;
+    _node_blockID[fe_elem->id()] = fe_elem->subdomain_id();
+  }
+
+  // search node neighbors and create other nodal data
+  createNodeHorizonBasedData();
+
+  // setup nodewise data for bond-associated deformation gradient
+  createBondAssocHorizonBasedData();
```

Comment:
Should this be only done if you're using the NOSPD model that uses this?

Reply:
Yes, this only applies to the NOSPD model. And since this is done only once during the mesh setup and it is not computationally very expensive, it is might be better not to introduce an additional parameter to check whether to execute this function or not.